Expand functionality test suite 4

fluiddot opened the pull request:
**Related PR:**
* https://github.com/WordPress/gutenberg/pull/52736

This PR expands the automated tests of [functionality test suite 4](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-suites/gutenberg/functionality-test-suites.md#test-suite-4) by including the following test cases:

- Spacer block - Dark mode - [TC004](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/darkmode.md#tc004)
- Buttons block - Dark mode - [TC005](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/darkmode.md#tc005)
- Group - Navigation up button works as expected - [TC006](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/group.md#tc006)
- Group - Navigation down works according to parent-first approach - [TC007](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/group.md#tc007)
- Group - Cross navigation between blocks works as expected - [TC008](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/group.md#tc008)
- Group - Check if in DarkMode all components gets proper colors - [TC009](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/group.md#tc009)

**To test:**
* Trigger optional UI tests job.
* Observe that all CI jobs succeed, including the full tests.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

fluiddot committed:
Add darkmode test cases for Spacer, Buttons and Group blocks

Apply fixes for Android

fluiddot committed:
Update Gutenberg ref

fluiddot committed:
Merge branch 'trunk' into add/expand-functionality-test-suite-4

# Conflicts:
#	gutenberg

fluiddot committed:
Add nested structure e2e test cases for Group block

fluiddot committed:
Add screenshots for nested structure e2e test cases

fluiddot committed:
Update Gutenberg ref

Update Gutenberg ref

fluiddot committed:
Remove unused imports from e2e test suite 4

fluiddot committed:
Update screenshots of Spacer block test cases

fluiddot committed:
Update nested structure e2e test cases for Group block

fluiddot committed:
Remove unused screenshots for Group block test cases

fluiddot commented on __device-tests__/gutenberg-editor-sanity-test-4-visual.test.js:
```diff
@@ -189,5 +231,169 @@ describe( 'Gutenberg Editor - Test Suite 4', () => {
+			// down works according to deepest-descendent-first approach.
+			const textXPath = isAndroid()
+				? `//android.widget.Button[@content-desc="${ blockNames.paragraph } Block. Row 1. Level ${ GROUP_NESTED_STRUCTURE_LEVELS }"]//android.widget.EditText`
+				: `(//*[@name="${ blockNames.paragraph } Block. Row 1. Level ${ GROUP_NESTED_STRUCTURE_LEVELS }"])[last()]`;
```
I tried using the same approach as on Android querying a text element (i.e. `XCUIElementTypeTextView`). However, this doesn't work for text elements being nested beyond two levels, for some reason they are not rendered as `XCUIElementTypeTextView`.

The approach I used is trying to get the deepest element that matches the label associated with the text element. In the future, it would be great to investigate why this happens.

fluiddot committed:
Update screenshot of Group block nested structure navigation test case

We need to wait until the floating toolbar is visible.

fluiddot commented:
Looks like one of the screenshots isn't matching on iOS e2e tests when running in CI 😭. I'd need to tweak the test so it can produce the exact screenshot consistently.

![image](https://github.com/wordpress-mobile/gutenberg-mobile/assets/14905380/df8864f9-5a4c-4680-87a1-df0e57555e9f)

dcalhoun commented:
> Looks like one of the screenshots isn't matching on iOS e2e tests when running in CI 😭. I'd need to tweak the test so it can produce the exact screenshot consistently.

I believe we generally dismiss the keyboard for any text-related screenshots for this reasons. However, doing so might negate assertions for correct Group block child selection. We may need to use a non-text-based block as the Group child.

fluiddot commented:
> I believe we generally dismiss the keyboard for any text-related screenshots for this reasons. However, doing so might negate assertions for correct Group block child selection. We may need to use a non-text-based block as the Group child.

Yep, good point, thanks! I'm going to try taking the screenshot without the keyboard, as we do in other cases asserting text elements. But using a non-text-base block will definitely resolve the issue 🙂.

fluiddot commented:
One of the issues I'm encountering in iOS is some kind of limitation on querying nested elements. Elements that exceed a certain nested level stop being accessible by Appium 🤔 .

fluiddot committed:
Update e2e test cases related to nested structure navigation

fluiddot commented:
@dcal

fluiddot commented:
@dcalhoun I finally solved the issue by using a different block structure. Instead of selecting a Paragraph block, I'm using a Spacer block, this way we avoid animated elements like the text caret. The PR is now ready for review, thanks 🙇 !

![image](https://github.com/wordpress-mobile/gutenberg-mobile/assets/14905380/dae39670-ab9b-476e-81e4-5f5d31366769)

dcalhoun commented on __device-tests__/gutenberg-editor-sanity-test-4-visual.test.js:
```diff
@@ -2,43 +2,85 @@
 describe( 'Gutenberg Editor - Test Suite 4', () => {
-	it( 'Spacer block - Spacer in horizontal layout works as expected', async () => {
-		await editorPage.addNewBlock( blockNames.spacer );
+	describe( 'Spacer block', () => {
```
💯 for leveraging `describe`.

dcalhoun commented on __device-tests__/gutenberg-editor-sanity-test-4-visual.test.js:
```diff
@@ -2,43 +2,85 @@
-	it( 'Spacer block - Spacer in horizontal layout works as expected', async () => {
-		await editorPage.addNewBlock( blockNames.spacer );
+	describe( 'Spacer block', () => {
+		it( 'Spacer in horizontal layout works as expected', async () => {
```
Sharing a few high-level testing thoughts for consideration, not necessary direct feedback on these changes. 

---

I would argue we should remove "Spacer" from individual test descriptions. My perception is that the `describe` and `it`/`test` functions are intended to be composed to make a descriptive assertion together. This suggestion would apply to any test description, not just the line referenced here. 

For example...

```js
describe("Spacer", () => {
  describe("when landscape orientation enabled", () => {
    it("should <meet an assertion>", () => {});
  });
});
```

Contrastingly, I do think it is worthwhile avoiding too much nesting (more than three levels) in test descriptions. So, in some contexts, it is worth combining context into a single describe, particularly if the context does not require multiple tests. 

```js
describe("Spacer", () => {
  it("should <meet an assertion> when landscape orientation enabled", () => {});
});
```

Additionally, when possible, we may benefit from communicating intent with as much detail in the test description as reasonably possible with consistent wording. There is obviously a balance between length and information, but there may be opportunity in a few of the current test descriptions. 

| Before | After |
| - | - |
| `Spacer in horizontal layout works as expected` | `should render proper alignment and spacing in landscape orientation` |
| `Check if in DarkMode all components gets proper colors` | `should render dark mode colors` | 
| `navigates nested structure` | `should traverse up and down nested blocks` |
| `Cross navigation between blocks works as expected` | `should navigate between nested blocks` |

fluiddot replied:
The current `it` description uses the original test case title we had in https://github.com/wordpress-mobile/test-cases.  Now that they are automated, I completely agree with your proposal to start using a more standard Jest test structure 👍. I'll update the PR with this these suggestions.

fluiddot replied:
Thinking twice, I'll update the test descriptions in a separate PR because it will require to re-generate all screenshots.

fluiddot replied:
Here is [the PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/6013) applying the above suggestions.

dcalhoun reviewed (approved):
Fantastic work! 👏🏻 

It looks like we need to address a test failure and merge conflict before merging. 

<details><summary>Test Failure</summary>

![Group block traversal visual snapshot diff](https://github.com/wordpress-mobile/gutenberg-mobile/assets/438664/7d9eff27-63e8-421e-b0ac-265c1d571242)


</details>

I left a few comments related to high-level testing approaches. Please feel free to merge this as-is.

fluiddot commented:
> It looks like we need to address a test failure and merge conflict before merging.
> 
> Test Failure

Yep, seems we need to wait until the floating toolbar is visible to take the screenshot. I'll update the tests.

fluiddot committed:
Increase wait time to ensure floating toolbar is displayed in nested structure e2e tests

The floating toolbar shows up animated, hence we need to wait extra time to ensure it's displayed before taking the screenshot.

fluiddot committed:
Update Gutenberg ref with merge commit

fluiddot committed:
Merge branch 'trunk' into add/expand-functionality-test-suite-4

# Conflicts:
#	gutenberg

fluiddot commented:
> > It looks like we need to address a test failure and merge conflict before merging.
> > Test Failure
> 
> Yep, seems we need to wait until the floating toolbar is visible to take the screenshot. I'll update the tests.

Fixed in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5977/commits/6c2aff149513a9d2743c45cf1dec0d338841e4db.
